File: solutions/107.binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        dfs(root, 0);
        return vector<vector<int>>(res.rbegin(), res.rend());
    }
private:
    vector<vector<int>> res;
    void dfs(TreeNode* root, int level) {
        if(!root) return;
        if(level == res.size()) res.push_back(vector<int>());
        res[level].push_back(root -> val);
        dfs(root -> left, level + 1);
        dfs(root -> right, level + 1);
    }
};

File: solutions/107.binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        if(root == null) return res;
        List<Integer> curRes = null;
        TreeNode curNode = null;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()) {
            int levelNum = queue.size();
            curRes = new ArrayList<>();
            for(int i = 0; i < levelNum; i++) {
                curNode = queue.poll();
                curRes.add(curNode.val);
                if(curNode.left != null) queue.offer(curNode.left);
                if(curNode.right != null) queue.offer(curNode.right);
            }
            res.add(0, curRes);
        }
        return res;
    }
}